kernel/src/client/default/parquet.rs

```diff
@@ -160,6 +180,69 @@ impl FileOpener for ParquetOpener {
     }
 }
 
+/// Implements [`FileOpener`] for a opening a parquet file from a presigned URL
+struct PresignedUrlOpener {
+    batch_size: usize,
+    limit: Option<usize>,
+    table_schema: SchemaRef,
+    client: reqwest::Client,
+}
+
+impl PresignedUrlOpener {
+    pub(crate) fn new(batch_size: usize, schema: SchemaRef) -> Self {
+        Self {
+            batch_size,
+            table_schema: schema,
+            limit: None,
+            client: reqwest::Client::new(),
+        }
+    }
+}
+
+impl FileOpener for PresignedUrlOpener {
+    fn open(&self, file_meta: FileMeta, _range: Option<Range<i64>>) -> DeltaResult<FileOpenFuture> {
+        let batch_size = self.batch_size;
+        let table_schema = self.table_schema.clone();
+        let limit = self.limit;
+        let client = self.client.clone(); // uses Arc internally according to reqwest docs
+
+        Ok(Box::pin(async move {
+            // fetch the file from the interweb
+            let reader = client.get(file_meta.location).send().await?.bytes().await?;
+            let metadata = ArrowReaderMetadata::load(&reader, Default::default())?;
+            let parquet_schema = metadata.schema();
+            let (indicies, requested_ordering) =
+                get_requested_indices(&table_schema, parquet_schema)?;
+
+            let options = ArrowReaderOptions::new();
+            let mut builder =
+                ParquetRecordBatchReaderBuilder::try_new_with_options(reader, options)?;
+            if let Some(mask) = generate_mask(
+                &table_schema,
+                parquet_schema,
+                builder.parquet_schema(),
+                &indicies,
+            ) {
+                builder = builder.with_projection(mask)
+            }
+
+            if let Some(limit) = limit {
+                builder = builder.with_limit(limit)
+            }
+
+            let reader = builder.with_batch_size(batch_size).build()?;
+
+            let stream = futures::stream::iter(reader);
+            let stream = stream.map(move |rbr| {
+                // re-order each batch if needed
+                rbr.map_err(Error::Arrow)
+                    .and_then(|rb| reorder_record_batch(rb, &requested_ordering))
+            });
+            Ok(stream.boxed())
```

Comment:
for the code above don't we do this elsewhere? can we reuse? maybe i'm missing something but i had the feeling this exists elsewhere

Reply:
yeah, it's similar to the code in the `ParquetOpener`. But that uses a `ParquetRecordBatchStreamBuilder` and this uses a `ParquetRecordBatchReaderBuilder`. They have a [common trait](https://arrow.apache.org/rust/parquet/arrow/arrow_reader/struct.ArrowReaderBuilder.html), but you'll notice that the `build` method is defined uniquely depending on the sync vs. async, and since we have both the types get really gross if you try and make this code shared.